README.md
Jimigrunge\Phalcon\Mailer
==============

Mailer wrapper over SwiftMailer for Phalcon.

## Installing ##

Install composer in a common location or in your project:

    curl -s http://getcomposer.org/installer | php

Create the composer.json file as follows:

```
{
	"require": {
		"jimigrunge/phalcon-mailer": "~3.0"
	}
}
```

Run the composer installer:

```bash
php composer.phar install
```

Add in your the code

    require_once('vendor/autoload.php');

## Configure ##

**SMTP**

```php
$config = [
    'driver' 	 => 'smtp',
    'host'	 	 => 'smtp.gmail.com',
    'port'	 	 => 465,
    'encryption' => 'ssl',
    'username'   => 'example@gmail.com',
    'password'	 => 'your_password',
    'from'		 => [
    		'email' => 'example@gmail.com',
    		'name'	=> 'YOUR FROM NAME'
    	]
];
```

> **Warning**
> If you are using GMAIL and get error `...Username and Password not accepted...` or `password incorrect`, 
> please using password token ([how get token?](https://support.google.com/accounts/answer/185833)) and fix in configuration
> 
> ```php
>   ...
>
>   'username' => 'example@gmail.com',
>   'password' => 'your_password_token',
>
>   ...
> ```

**Sendmail**

```php
$config = [
    'driver' 	 => 'sendmail',
	'sendmail' 	 => '/usr/sbin/sendmail -bs',
    'from'		 => [
    	'email' => 'example@gmail.com',
    	'name'	=> 'YOUR FROM NAME'
    ]
];
```

## Example ##

### createMessage() ###

```php
$mailer = new \Phalcon\Ext\Mailer\Manager($config);

$message = $mailer->createMessage()
		->to('example_to@gmail.com', 'OPTIONAL NAME')
		->subject('Hello world!')
		->content('Hello world!');

// Set the Cc addresses of this message.
$message->cc('example_cc@gmail.com');

// Set the Bcc addresses of this message.
$message->bcc('example_bcc@gmail.com');

// Send message
$message->send();
```

### createMessageFromView() ###

> **Warning**
> If you want to use as a template engine VOLT (.volt), 
> please setup "view" service according to the official [docs](http://docs.phalconphp.com/en/latest/reference/volt.html#activating-volt) Phalcon

```php
/**
 * Global viewsDir for current instance Mailer\Manager.
 * 
 * This parameter is OPTIONAL, If it is not specified, 
 * use DI from "view" service (getViewsDir)
 */
$config['viewsDir'] = __DIR__ . '/views/email/';

$mailer = new \Phalcon\Ext\Mailer\Manager($config);

// view relative to the folder viewsDir (REQUIRED)
$viewPath = 'email/example_message';

// Set variables to views (OPTIONAL)
$params [ 
	'var1' => 'VAR VALUE 1',
	'var2' => 'VAR VALUE 2',
	...
	'varN' => 'VAR VALUE N',
];

/**
 * The local path to the folder viewsDir only this message. (OPTIONAL)
 * 
 * This parameter is OPTIONAL, If it is not specified, 
 * use global parameter "viewsDir" from configuration.
 */
$viewsDirLocal = __DIR__ . '/views/email/local/';


$message = $mailer->createMessageFromView($viewPath, $params, $viewsDirLocal)
		->to('example_to@gmail.com', 'OPTIONAL NAME')
		->subject('Hello world!');

// Set the Cc addresses of this message.
$message->cc('example_cc@gmail.com');

// Set the Bcc addresses of this message.
$message->bcc('example_bcc@gmail.com');

// Send message
$message->send();
```

## Events ##
- mailer:beforeCreateMessage
- mailer:afterCreateMessage
- mailer:beforeSend
- mailer:afterSend
- mailer:beforeAttachFile
- mailer:afterAttachFile


CHANGELOG.md
# CHANGELOG

## v3.0.0 (2017-09-29)
- Updated Swiftmailer to version 6
- Removed functionality dependent on removed Swift_MailTransport class

## v2.0.3 (2015-12-18)
- Added method `contentAlternative()` - optionally an alternative text body 
- fix [#5](https://github.com/phalcon-ext/mailer/issues/5)

## v2.0.2 (2015-03-03)
- fix [#2](https://github.com/phalcon-ext/mailer/issues/2)

## v2.0.1 (2014-11-05)
- Changes in the private logic...
  - Added protected method `isInitSwiftMailer()`
  - Added lazy init `SwiftMailer`
  - Added check for the required of dependency injection object
  
## v2.0.0 (2014-10-16)
- rename root namespace `\Phalcon\Mailer` to `\Phalcon\Ext\Mailer`
- remove deprecated method `getMessage()` in class `\Phalcon\Ext\Mailer\Message`

## v1.1.0 (2014-10-16)
- added method `getSwiftMessage()` in class `\Phalcon\Mailer\Message` analogue for `getMessage()`
- method `getMessage()` in class `\Phalcon\Mailer\Message` is deprecated (will removed since 2.0.0)

## v1.0.2 (2014-09-17)
- fixed bug when you create a message via View (incorrect namespace)

## v1.0.1 (2014-09-15)
- fixed bug in the attachment file.
- In `\Phalcon\Mailer\Message`, creation SwiftMailer of instances moved to the DI

## v1.0.0 (2014-09-14)
- Release of component :) 


src/Message.php
<?php
/**
 * Message.php 2014-08-31 04:50
 * ----------------------------------------------
 *
 *
 * @author      Stanislav Kiryukhin <korsar.zn@gmail.com>
 * @copyright   Copyright (c) 2014, CKGroup.ru
 *
 * ----------------------------------------------
 * All Rights Reserved.
 * ----------------------------------------------
 */
namespace Jimigrunge\Phalcon\Mailer;

/**
 * Class Message
 */
class Message
{
    /**
     * content type of PLAIN text.
     */
    const CONTENT_TYPE_PLAIN = 'text/plain';

    /**
     * content type HTML text.
     */
    const CONTENT_TYPE_HTML = 'text/html';

    /**
     * @var \Phalcon\Ext\Mailer\Manager
     */
    protected $manager;

    /**
     * @var \Swift_Message
     */
    protected $swiftMessage;

    /**
     * An array of email which failed send to recipients.
     *
     * @var array
     */
    protected $failedRecipients = [];

    /**
     * Create a new Message using $mailer for sending from SwiftMailer
     *
     * @param Manager $manager
     */
    public function __construct(Manager $manager)
    {
        $this->manager = $manager;
    }

    /**
     * Set the from address of this message.
     *
     * You may pass an array of addresses if this message is from multiple people.
     * Example: array('receiver@domain.org', 'other@domain.org' => 'A name')
     *
     * If $name is passed and the first parameter is a string, this name will be
     * associated with the address.
     *
     * @param string|array $email
     * @param string|null $name optional
     *
     * @return $this
     *
     * @see \Swift_Message::setFrom()
     */
    public function from($email, $name = null)
    {
        $email = $this->normalizeEmail($email);
        $this->getSwiftMessage()->setFrom($email, $name);

        return $this;
    }

    /**
     * Get the from address of this message.
     *
     * @return string
     *
     * @see \Swift_Message::getFrom()
     */
    public function getFrom()
    {
        return $this->getSwiftMessage()->getFrom();
    }

    /**
     * Set the reply-to address of this message.
     *
     * You may pass an array of addresses if replies will go to multiple people.
     * Example: array('receiver@domain.org', 'other@domain.org' => 'A name')
     *
     * If $name is passed and the first parameter is a string, this name will be
     * associated with the address.
     *
     * @param string|array $email
     * @param string|null $name optional
     *
     * @return $this
     *
     * @see \Swift_Message::setReplyTo()
     */
    public function replyTo($email, $name = null)
    {
        $email = $this->normalizeEmail($email);
        $this->getSwiftMessage()->setReplyTo($email, $name);

        return $this;
    }

    /**
     * Get the reply-to address of this message.
     *
     * @return string
     *
     * @see \Swift_Message::getReplyTo()
     */
    public function getReplyTo()
    {
        return $this->getSwiftMessage()->getReplyTo();
    }

    /**
     * Set the to addresses of this message.
     *
     * If multiple recipients will receive the message an array should be used.
     * Example: array('receiver@domain.org', 'other@domain.org' => 'A name')
     *
     * If $name is passed and the first parameter is a string, this name will be
     * associated with the address.
     *
     * @param string|array $email
     * @param string|null $name optional
     *
     * @return $this
     *
     * @see \Swift_Message::setTo()
     */
    public function to($email, $name = null)
    {
        $email = $this->normalizeEmail($email);
        $this->getSwiftMessage()->setTo($email, $name);

        return $this;
    }

    /**
     * Get the To addresses of this message.
     *
     * @return array
     *
     * @see \Swift_Message::getTo()
     */
    public function getTo()
    {
        return $this->getSwiftMessage()->getTo();
    }

    /**
     * Set the Cc addresses of this message.
     *
     * If multiple recipients will receive the message an array should be used.
     * Example: array('receiver@domain.org', 'other@domain.org' => 'A name')
     *
     * If $name is passed and the first parameter is a string, this name will be
     * associated with the address.
     *
     * @param string|array $email
     * @param string|null $name optional
     *
     * @return $this
     *
     * @see \Swift_Message::setCc()
     */
    public function cc($email, $name = null)
    {
        $email = $this->normalizeEmail($email);
        $this->getSwiftMessage()->setCc($email, $name);

        return $this;
    }

    /**
     * Get the Cc address of this message.
     *
     * @return array
     *
     * @see \Swift_Message::getCc()
     */
    public function getCc()
    {
        return $this->getSwiftMessage()->getCc();
    }

    /**
     * Set the Bcc addresses of this message.
     *
     * If multiple recipients will receive the message an array should be used.
     * Example: array('receiver@domain.org', 'other@domain.org' => 'A name')
     *
     * If $name is passed and the first parameter is a string, this name will be
     * associated with the address.
     *
     * @param string|array $email
     * @param string|null $name optional
     *
     * @return $this
     *
     * @see \Swift_Message::setBcc()
     */
    public function bcc($email, $name = null)
    {
        $email = $this->normalizeEmail($email);
        $this->getSwiftMessage()->setBcc($email, $name);

        return $this;
    }

    /**
     * Get the Bcc addresses of this message.
     *
     * @return array
     *
     * @see \Swift_Message::getBcc()
     */
    public function getBcc()
    {
        return $this->getSwiftMessage()->getBcc();
    }

    /**
     * Set the sender of this message.
     *
     * This does not override the From field, but it has a higher significance.
     *
     * @param string|array $email
     * @param string|null $name optional
     *
     * @return $this
     *
     * @see \Swift_Message::setSender()
     */
    public function sender($email, $name = null)
    {
        $email = $this->normalizeEmail($email);
        $this->getSwiftMessage()->setSender($email, $name);

        return $this;
    }

    /**
     * Get the sender of this message.
     *
     * @return string
     *
     * @see \Swift_Message::getSender()
     */
    public function getSender()
    {
        return $this->getSwiftMessage()->getSender();
    }

    /**
     * Set the subject of this message.
     *
     * @param string $subject
     *
     * @return $this
     *
     * @see \Swift_Message::setSubject()
     */
    public function subject($subject)
    {
        $this->getSwiftMessage()->setSubject($subject);

        return $this;
    }

    /**
     * Get the subject of this message.
     *
     * @return string
     *
     * @see \Swift_Message::getSubject()
     */
    public function getSubject()
    {
        return $this->getSwiftMessage()->getSubject();
    }

    /**
     * Set the body of this message, either as a string, or as an instance of
     * {@link \Swift_OutputByteStream}.
     *
     * @param mixed $content
     * @param string $contentType optional
     * @param string $charset optional
     *
     * @return $this
     *
     * @see \Swift_Message::setBody()
     */
    public function content($content, $contentType = self::CONTENT_TYPE_HTML, $charset = null)
    {
        $this->getSwiftMessage()->setBody($content, $contentType, $charset);

        return $this;
    }

    /**
     * Get the body of this message as a string.
     *
     * @return string
     *
     * @see \Swift_Message::getBody()
     */
    public function getContent()
    {
        return $this->getSwiftMessage()->getBody();
    }

    /**
     * Add optionally an alternative body
     *
     * @param string $content
     * @param string $contentType optional
     * @param string $charset optional
     *
     * @return $this
     */
    public function contentAlternative($content, $contentType = null, $charset = null)
    {
        $this->getSwiftMessage()->addPart($content, $contentType, $charset);

        return $this;
    }

    /**
     * Set the Content-type of this message.
     *
     * @param string $contentType
     *
     * @return $this
     *
     * @see \Swift_Message::setContentType()
     */
    public function contentType($contentType)
    {
        $this->getSwiftMessage()->setContentType($contentType);

        return $this;
    }

    /**
     * Get the Content-type of this message.
     *
     * @return string
     *
     * @see \Swift_Message::getContentType()
     */
    public function getContentType()
    {
        return $this->getSwiftMessage()->getContentType();
    }

    /**
     * Set the character set of this message.
     *
     * @param string $charset
     *
     * @return $this
     *
     * @see \Swift_Message::setCharset()
     */
    public function charset($charset)
    {
        $this->getSwiftMessage()->setCharset($charset);

        return $this;
    }

    /**
     * Get the character set of this message.
     *
     * @return string
     *
     * @see \Swift_Message::getCharset()
     */
    public function getCharset()
    {
        return $this->getSwiftMessage()->getCharset();
    }

    /**
     * Set the priority of this message.
     *
     * The value is an integer where 1 is the highest priority and 5 is the lowest.
     *
     * @param int $priority
     *
     * @return $this
     *
     * @see \Swift_Message::setPriority()
     */
    public function priority($priority)
    {
        $this->getSwiftMessage()->setPriority($priority);

        return $this;
    }

    /**
     * Get the priority of this message.
     *
     * The returned value is an integer where 1 is the highest priority and 5
     * is the lowest.
     *
     * @return int
     *
     * @see \Swift_Message::getPriority()
     */
    public function getPriority()
    {
        return $this->getSwiftMessage()->getPriority();
    }

    /**
     * Ask for a delivery receipt from the recipient to be sent to $addresses
     *
     * @param array $email
     *
     * @return $this
     *
     * @see \Swift_Message::setReadReceiptTo()
     */
    public function setReadReceiptTo($email)
    {
        $email = $this->normalizeEmail($email);
        $this->getSwiftMessage()->setReadReceiptTo($email);

        return $this;
    }

    /**
     * An array of email which failed send to recipients.
     *
     * @return array
     */
    public function getFailedRecipients()
    {
        return $this->failedRecipients;
    }

    /**
     * Get the addresses to which a read-receipt will be sent.
     *
     * @return string
     *
     * @see \Swift_Message::getReadReceiptTo()
     */
    public function getReadReceiptTo()
    {
        return $this->getSwiftMessage()->getReadReceiptTo();
    }

    /**
     * Set the return-path (the bounce address) of this message.
     *
     * @param string $email
     *
     * @return $this
     *
     * @see \Swift_Message::setReturnPath()
     */
    public function setReturnPath($email)
    {
        $this->getSwiftMessage()->setReturnPath($email);

        return $this;
    }

    /**
     * Get the return-path (bounce address) of this message.
     *
     * @return string
     *
     * @see \Swift_Message::getReturnPath()
     */
    public function getReturnPath()
    {
        return $this->getSwiftMessage()->getReturnPath();
    }

    /**
     * Set the format of this message (flowed or fixed).
     *
     * @param string $format
     *
     * @return string
     *
     * @see \Swift_Message::setFormat()
     */
    public function setFormat($format)
    {
        $this->getSwiftMessage()->setFormat($format);

        return $this;
    }

    /**
     * Get the format of this message (i.e. flowed or fixed).
     *
     * @return string
     *
     * @see \Swift_Message::getFormat()
     */
    public function getFormat()
    {
        return $this->getSwiftMessage()->getFormat();
    }

    /**
     * Attach a file to the message.
     *
     * Events:
     * - mailer:beforeAttachFile
     * - mailer:afterAttachFile
     *
     * @param  string $file
     * @param  array $options optional
     *
     * @return $this
     *
     * @see Phalcon\Ext\Mailer\Message::createAttachmentViaPath()
     * @see Phalcon\Ext\Mailer\Message::prepareAttachment()
     */
    public function attachment($file, array $options = [])
    {
        $attachment = $this->createAttachmentViaPath($file);

        return $this->prepareAttachment($attachment, $options);
    }

    /**
     * Attach in-memory data as an attachment.
     *
     * @param  string $data
     * @param  string $name
     * @param  array $options optional
     *
     * @return Message
     *
     * @see Phalcon\Ext\Mailer\Message::createAttachmentViaData()
     * @see Phalcon\Ext\Mailer\Message::prepareAttachment()
     */
    public function attachmentData($data, $name, array $options = [])
    {
        $attachment = $this->createAttachmentViaData($data, $name);

        return $this->prepareAttachment($attachment, $options);
    }

    /**
     * Embed a file in the message and get the CID.
     *
     * @param  string $file
     *
     * @return string
     */
    public function embed($file)
    {
        $embed = $this->createEmbedViaPath($file);

        return $this->getSwiftMessage()->embed($embed);
    }

    /**
     * Embed in-memory data in the message and get the CID.
     *
     * @param  string $data
     * @param  string $name
     * @param  string $contentType
     *
     * @return string
     */
    public function embedData($data, $name, $contentType = null)
    {
        $embed = $this->createEmbedViaData($data, $name, $contentType);

        return $this->getSwiftMessage()->embed($embed);
    }

    /**
     * Return a {@link \Swift_Message} instance
     *
     * @return \Swift_Message
     */
    public function getSwiftMessage()
    {
        if (!$this->swiftMessage) {
            $this->swiftMessage = $this->getManager()->getSwift()->createMessage();
        }

        return $this->swiftMessage;
    }

    /**
     * Return a {@link \Phalcon\Ext\Mailer\Manager} instance
     *
     * @return \Phalcon\Ext\Mailer\Manager
     */
    public function getManager()
    {
        return $this->manager;
    }

    /**
     * Send the given Message like it would be sent in a mail client.
     *
     * All recipients (with the exception of Bcc) will be able to see the other
     * recipients this message was sent to.
     *
     * Recipient/sender data will be retrieved from the Message object.
     *
     * The return value is the number of recipients who were accepted for
     * delivery.
     *
     * Events:
     * - mailer:beforeSend
     * - mailer:afterSend
     *
     * @return int
     *
     * @see \Swift_Mailer::send()
     */
    public function send()
    {
        $eventManager = $this->getManager()->getEventsManager();

        if ($eventManager) {
            $result = $eventManager->fire('mailer:beforeSend', $this);
        } else {
            $result = true;
        }

        if ($result !== false) {

            $this->failedRecipients = [];

            $count = $this->getManager()->getSwift()->send($this->getSwiftMessage(), $this->failedRecipients);

            if ($eventManager) {
                $eventManager->fire('mailer:afterSend', $this, [$count, $this->failedRecipients]);
            }

            return $count;

        } else {
            return false;
        }
    }

    /**
     * Prepare and attach the given attachment.
     *
     * @param  \Swift_Attachment $attachment
     * @param  array $options optional
     *
     * @return $this
     *
     * @see \Swift_Message::attach()
     */
    protected function prepareAttachment(\Swift_Attachment $attachment, array $options = [])
    {
        if (isset($options['mime'])) {
            $attachment->setContentType($options['mime']);
        }

        if (isset($options['as'])) {
            $attachment->setFilename($options['as']);
        }

        $eventManager = $this->getManager()->getEventsManager();

        if ($eventManager) {
            $result = $eventManager->fire('mailer:beforeAttachFile', $this, [$attachment]);
        } else {
            $result = true;
        }

        if ($result !== false) {
            $this->getSwiftMessage()->attach($attachment);

            if ($eventManager) {
                $eventManager->fire('mailer:afterAttachFile', $this, [$attachment]);
            }
        }

        return $this;
    }

    /**
     * Create a Swift new Attachment from a filesystem path.
     *
     * @param   string $file
     *
     * @return \Swift_Attachment
     *
     * @see \Swift_Attachment::fromPath()
     */
    protected function createAttachmentViaPath($file)
    {
        /** @var $byteStream \Swift_ByteStream_FileByteStream */
        $byteStream = $this->getManager()->getDI()->get('\Swift_ByteStream_FileByteStream', [$file]);

        /** @var $image \Swift_Attachment */
        $attachment = $this->getManager()->getDI()->get('\Swift_Attachment')
            ->setFile($byteStream);

        return $attachment;
    }

    /**
     * Create a Swift Attachment instance from data.
     *
     * @param string $data
     * @param string $name optional
     *
     * @return \Swift_Attachment
     *
     * @see \Swift_Attachment::newInstance()
     */
    protected function createAttachmentViaData($data, $name)
    {
        return $this->getManager()->getDI()->get('\Swift_Attachment', [$data, $name]);
    }

    /**
     * Create a Swift new Image from a filesystem path.
     *
     * @param string $file
     *
     * @return \Swift_Image
     *
     * @see \Swift_Image::fromPath()
     */
    protected function createEmbedViaPath($file)
    {
        /** @var $byteStream \Swift_ByteStream_FileByteStream */
        $byteStream = $this->getManager()->getDI()->get('\Swift_ByteStream_FileByteStream', [$file]);

        /** @var $image \Swift_Image */
        $image = $this->getManager()->getDI()->get('\Swift_Image')
            ->setFile($byteStream);

        return $image;
    }

    /**
     * Create a Swift new Image.
     *
     * @param string $data
     * @param string|null $filename
     * @param string|null $contentType
     *
     * @return \Swift_Image
     *
     * @see \Swift_Image::newInstance()
     */
    protected function createEmbedViaData($data, $filename = null, $contentType = null)
    {
        return $this->getManager()->getDI()->get('\Swift_Image', [$data, $filename, $contentType]);
    }

    /**
     * Normalize IDN domains.
     *
     * @param $email
     *
     * @return array|string
     */
    protected function normalizeEmail($email)
    {
        if (is_array($email)) {

            $emails = [];

            foreach ($email as $k => $v) {
                if (is_int($k)) {
                    $emails[$k] = $this->getManager()->normalizeEmail($v);
                } else {
                    $k = $this->getManager()->normalizeEmail($k);
                    $emails[$k] = $v;
                }
            }

            return $emails;

        } else {
            return $this->getManager()->normalizeEmail($email);
        }
    }
}


src/Manager.php
<?php
/**
 * Manager.php 2014-08-31 04:11
 * ----------------------------------------------
 *
 * @author      Stanislav Kiryukhin <korsar.zn@gmail.com>
 * @copyright   Copyright (c) 2014, CKGroup.ru
 *
 * ----------------------------------------------
 * All Rights Reserved.
 * ----------------------------------------------
 */
namespace Jimigrunge\Phalcon\Mailer;

use Phalcon\Config;
use Phalcon\Mvc\User\Component;
use Phalcon\Mvc\View;
use Phalcon\DiInterface;

/**
 * Class Manager
 */
class Manager extends Component
{
    const AUTHENTICATION_MODE_CRAM_MD5  = 'CRAM-MD5';
    const AUTHENTICATION_MODE_LOGIN     = 'LOGIN';
    const AUTHENTICATION_MODE_NTLM      = 'NTLM';
    const AUTHENTICATION_MODE_PLAIN     = 'PLAIN';
    const AUTHENTICATION_MODE_OAUTH     = 'XOAUTH2';

    const ENCRYPTION_NONE = '';
    const ENCRYPTION_SSL  = 'ssl';
    const ENCRYPTION_TLS  = 'tls';

    /**
     * @var array
     */
    protected $config = [];

    /**
     * @var \Swift_Transport
     */
    protected $transport;

    /**
     * @var \Swift_Mailer
     */
    protected $mailer;

    /**
     * @var \Phalcon\Mvc\View\Simple
     */
    protected $view;

    /**
     * Create a new MailerManager component using $config for configuring
     *
     * @param array $config
     */
    public function __construct(array $config)
    {
        $this->configure($config);
    }

    /**
     * Create a new Message instance.
     *
     * Events:
     * - mailer:beforeCreateMessage
     * - mailer:afterCreateMessage
     *
     * @return \Phalcon\Ext\Mailer\Message
     */
    public function createMessage()
    {
        $eventsManager = $this->getEventsManager();

        if ($eventsManager) {
            $eventsManager->fire('mailer:beforeCreateMessage', $this);
        }

        /** @var $message Message */
        $message = $this->getDI()->get('\Phalcon\Ext\Mailer\Message', [$this]);

        if (($from = $this->getConfig('from'))) {
            $message->from($from['email'], isset($from['name']) ? $from['name'] : null);
        }

        if ($eventsManager) {
            $eventsManager->fire('mailer:afterCreateMessage', $this, [$message]);
        }

        return $message;
    }

    /**
     * Create a new Message instance.
     * For the body of the message uses the result of render of view
     *
     * Events:
     * - mailer:beforeCreateMessage
     * - mailer:afterCreateMessage
     *
     * @param string $view
     * @param array $params optional
     * @param null|string $viewsDir optional
     *
     * @return \Phalcon\Ext\Mailer\Message
     *
     * @see \Phalcon\Ext\Mailer\Manager::createMessage()
     */
    public function createMessageFromView($view, $params = [], $viewsDir = null)
    {
        $message = $this->createMessage();
        $message->content($this->renderView($view, $params, $viewsDir), $message::CONTENT_TYPE_HTML);

        return $message;
    }

    /**
     * Return a {@link \Swift_Mailer} instance
     *
     * @return \Swift_Mailer
     */
    public function getSwift()
    {
        if (!$this->isInitSwiftMailer()) {
            $this->registerSwiftMailer();
        }

        return $this->mailer;
    }

    /**
     * Normalize IDN domains.
     *
     * @param $email
     *
     * @return string
     *
     * @see \Phalcon\Ext\Mailer\Manager::punycode()
     */
    public function normalizeEmail($email)
    {
        if (preg_match('#[^(\x20-\x7F)]+#', $email)) {

            list($user, $domain) = explode('@', $email);

            return $user . '@' . $this->punycode($domain);

        } else {
            return $email;
        }
    }

    /**
     * Configure MailerManager class
     *
     * @param array $config
     *
     * @see \Phalcon\Ext\Mailer\Manager::registerSwiftTransport()
     * @see \Phalcon\Ext\Mailer\Manager::registerSwiftMailer()
     */
    protected function configure(array $config)
    {
        $this->config = $config;
    }

    /**
     * Create a new Driver-mail of SwiftTransport instance.
     *
     * Supported driver-mail:
     * - smtp
     * - sendmail
     *
     */
    protected function registerSwiftTransport()
    {
        switch ($driver = $this->getConfig('driver')) {
            case 'smtp':
                $this->transport = $this->registerTransportSmtp();
                break;

            case 'sendmail':
                $this->transport = $this->registerTransportSendmail();
                break;

            default:
                throw new \InvalidArgumentException(sprintf('Driver-mail "%s" is not supported', $driver));
        }
    }

    /**
     * Create a new SmtpTransport instance.
     *
     * @return \Swift_SmtpTransport
     *
     * @see \Swift_SmtpTransport
     */
    protected function registerTransportSmtp()
    {
        $config = $this->getConfig();

        /** @var $transport \Swift_SmtpTransport: */
        $transport = $this->getDI()->get('\Swift_SmtpTransport');

        if (isset($config['encryption'])) {
            $transport->setEncryption($config['encryption']);
        }

        if (isset($config['host'])) {
            $transport->setHost($config['host']);
        } else {
            $transport->setHost('localhost');
        }

        if (isset($config['port'])) {
            $transport->setPort($config['port']);
        } else {
            switch ($transport->getEncryption()) {
                case static::ENCRYPTION_SSL:
                    $transport->setPort(465);
                    break;

                case static::ENCRYPTION_TLS:
                    $transport->setPort(587);
                    break;

                default:
                    $transport->setPort(25);
            }
        }

        if (isset($config['auth_mode'])) {
            $transport->setAuthMode($config['auth_mode']);
        }

        if (isset($config['username'])) {
            $transport->setUsername($this->normalizeEmail($config['username']));
        }

        if (isset($config['password'])) {
            $transport->setPassword($config['password']);
        }

        return $transport;
    }

    /**
     * Get option config or the entire array of config, if the parameter $key is not specified.
     *
     * @param null $key
     * @param null $default
     *
     * @return string|array
     */
    protected function getConfig($key = null, $default = null)
    {
        if ($key !== null) {
            if (isset($this->config[$key])) {
                return $this->config[$key];
            } else {
                return $default;
            }

        } else {
            return $this->config;
        }
    }

    /**
     * Convert UTF-8 encoded domain name to ASCII
     *
     * @param $str
     *
     * @return string
     */
    protected function punycode($str)
    {
        if (function_exists('idn_to_ascii')) {
            return idn_to_ascii($str);
        } else {
            return $str;
        }
    }

    /**
     * Create a new SendmailTransport instance.
     *
     * @return \Swift_SendmailTransport
     *
     * @see \Swift_SendmailTransport
     */
    protected function registerTransportSendmail()
    {
        /** @var $transport \Swift_SendmailTransport */
        $transport = $this->getDI()->get('\Swift_SendmailTransport')
            ->setCommand($this->getConfig('sendmail', '/usr/sbin/sendmail -bs'));

        return $transport;
    }

    /**
     * Register SwiftMailer
     *
     * @see \Swift_Mailer
     */
    protected function registerSwiftMailer()
    {
        $this->registerSwiftTransport();
        $this->mailer = $this->getDI()->get('\Swift_Mailer', [$this->transport]);
    }

    /**
     * Renders a view
     *
     * @param $viewPath
     * @param $params
     * @param null $viewsDir
     *
     * @return string
     */
    protected function renderView($viewPath, $params, $viewsDir = null)
    {
        $view = $this->getView();

        if ($viewsDir !== null) {
            $view = clone $view;
            $view->setViewsDir($viewsDir);
        }

        if (!$view->getViewsDir()) {
            throw new \RuntimeException('Please set config property "viewsDir"');
        }

        return $view->render($viewPath, $params);
    }

    /**
     * Return a {@link \Phalcon\Mvc\View\Simple} instance
     *
     * @return \Phalcon\Mvc\View\Simple
     */
    protected function getView()
    {
        if ($this->view) {
            return $this->view;
        } else {

            /** @var $viewApp \Phalcon\Mvc\View */
            $viewApp = $this->getDI()->get('view');

            if (!($viewsDir = $this->getConfig('viewsDir'))) {
                $viewsDir = $viewApp->getViewsDir();
            }

            /** @var $view \Phalcon\Mvc\View\Simple */
            $view = $this->getDI()->get('\Phalcon\Mvc\View\Simple');
            $view->setViewsDir($viewsDir);

            if ($engines = $viewApp->getRegisteredEngines()) {
                $view->registerEngines($engines);
            }

            return $this->view = $view;
        }
    }

    /**
     * Check init SwiftMailer
     *
     * @return bool
     */
    protected function isInitSwiftMailer()
    {
        return $this->mailer && $this->transport;
    }

    /**
     * Returns the internal dependency injector
     *
     * @return \Phalcon\DiInterface
     */
    public function getDI()
    {
        if (!($di = parent::getDI()) && !($di instanceof DiInterface)) {
            throw new \RuntimeException('A dependency injection object is required to access internal services');
        }

        return $di;
    }
}
